truss/templates/server/model_wrapper.py

```diff
@@ -39,28 +100,39 @@ def load(self):
         if _signature_accepts_keyword_arg(model_class_signature, "secrets"):
             model_init_params["secrets"] = SecretsResolver.get_secrets(self._config)
         self._model = model_class(**model_init_params)
+
         if hasattr(self._model, "load"):
             self._model.load()
-        self.ready = True
 
-    def preprocess(self, request: Dict) -> Dict:
+    async def preprocess(
+        self,
+        payload: Union[Dict, CloudEvent, ModelInferRequest],
+        headers: Dict[str, str] = None,
+    ) -> Union[Dict, ModelInferRequest]:
         if not hasattr(self._model, "preprocess"):
-            return request
-        return self._model.preprocess(request)
+            return payload
+        return self._model.preprocess(payload)
 
-    def postprocess(self, request: Dict) -> Dict:
+    def postprocess(
+        self, response: Union[Dict, ModelInferResponse], headers: Dict[str, str] = None
+    ) -> Dict:
         if not hasattr(self._model, "postprocess"):
-            return request
-        return self._model.postprocess(request)
+            return response
+        return self._model.postprocess(response)
 
-    def predict(self, request: Dict) -> Dict[str, List]:
-        response = {}
+    def predict(
+        self, payload: Union[Dict, ModelInferRequest], headers: Dict[str, str] = None
+    ) -> Union[Dict, ModelInferResponse]:
         try:
-            return self._model.predict(request)
+            self._predict_lock.acquire()
```

Comment:
I'm confused, does this locking not limit concurrent processing. It may be worth putting a code comment to explain. What's the purpose of this lock?

Reply:
It does for the `predict` function, which is completely compute bound.   Context switching during this function would not gain us any benefit and only slow down due to context switching and/or fail, as in the GPU case.

What this does though is release the CPU to work on I/O operations immediately after the GPU has finished with its tasks and another request can instantly jump in to work.  I hypothesize not being able to do this is what causes our big swings in GPU util down to zero and back to 100% utilization.

Reply:
There is more optimization here for multi-core workloads that can be done, but this is the first step.

Reply:
There are possible models that just proxy to an external service. e.g. [this](https://github.com/basetenlabs/truss/blob/main/examples/invoice-parser/model/model.py#L18). This lock will drastically reduce performance for that kind of models. I wonder if there's another way of tackling the GPU utilization issue that doesn't affect this case.

Reply:
Without an API that allows for that context, it isn't possible. In newer versions we could move towards an async API (with a defined interface for models), which would allow this context to be passed back to the container.  For example, having `predict` be async would allow the model to behave as such: 

```
await extract_pdf_data(url, self._secrets)
```

Worth noting that this PR won't affect performance of these models that call out to other services as the concurrency would still reduce to 1, as it is now.

Even in the case we have these models allow for IO, this still causes potential memory corruption issues for models that write into the same blocks of GPU memory, as Bola indicated in this PR.

My overall recommendation for now is to put this code into preprocess, as that will be handled asynchronously with this change.

Reply:
Thanks for explaining. It makes sense to go with this for now. In the future, when we support the async interface on predict, we could optimize for the external service proxy case better.